Edx Oauth2 provider code flow and tests.

dominicfollett opened the pull request.

marcoow commented on README.md:
```diff
@@ -156,14 +156,37 @@ RailsApiAuth.tap do |raa|
+headers = {
+	username: "alice", 
+	auth_code: "alices_authorization_code", 
+	grant_type: "edx_auth_code"
```
never use tabs please

dominicfollett replied:
Noted, will update now.

marcoow commented on README.md:
```diff
@@ -156,14 +156,37 @@ RailsApiAuth.tap do |raa|
+It is nesescary to include the Edx username in the when making a call
+rails_api_auth call /token. When rails_api_auth interfaces with Edx's
+user api, the username is need to retrieve user data, not just a valid
+oauth2 token. 
```
This sounds strange and if I'm not mistaken doesn't conform to the spec. Can you explain why that is?

dominicfollett replied:
It is odd, however the Edx [user api](http://edx.readthedocs.io/projects/edx-platform-api/en/latest/user/overview.html#user-api-resources-and-endpoints) specified that a username is needed to retrieve user profile data. I'm not sure why they handled it this way. It certainly makes it less wieldy than the Google or Facebook method.

dominicfollett commented:
@marcoow I notice the CI builds are failing due to a dependency error:

```bash
An error occurred while installing ruby_dep (1.2.0), and Bundler cannot continue.
Make sure that `gem install ruby_dep -v '1.2.0'` succeeds before bundling.
```

I'm not sure how this might relate to my PR?

marcoow commented:
@dominicfollett: rebase on `master` - I fixed the tests in #43

dominicfollett commented:
@marcoow I'm getting test fails, due to a username being needed with Edx user API calls. I can fix this by providing my own tests to exercise this, I just wanted to check if you were ok with this, or perhaps want me to find another to get around this? I don't want to bloat your testing suite without due cause.

marcoow commented:
@dominicfollett: can you point me to the change you're referring to?

dominicfollett commented:
@marcoow I've added shared_examples/oauth2_edx_shared_requests.rb, where username needs to be set so it is passed as a parameter with the /token call. I also added a test to exercise the case where a username is not supplied.

marcoow commented:
Can you revert all those file mode changes (644 -> 755)? I don't think those are correct.

dominicfollett commented:
@marcoow I've updated that mode change. Apologies!

marcoow commented on .travis.yml:
```diff
@@ -1,10 +1,9 @@
-  - rbx-2
+  - 2.2.4
+  - 2.3.0
+  - jruby-9.0.5.0
```
this seems unrelated

marcoow commented on app/controllers/oauth2_controller.rb:
```diff
@@ -16,6 +16,15 @@ def create
     when 'google_auth_code'
       authenticate_with_google(params[:auth_code])
     else
+      other_authenticators
```
Why do we handle edx and unknown authenticators in a separate method?

dominicfollett replied:
The Travis CI complains about method lengths longer than 10 lines.

marcoow replied:
Can you disable that check for this specific method. Splitting it up like this might make Rubocop happy but actually makes the code much worse ;)

dominicfollett replied:
Ok sure thing :)

marcoow commented on gemfiles/rails_3.gemfile:
```diff
@@ -2,7 +2,7 @@
 source "https://rubygems.org"
 
-gem "rails", "3.2.22"
+gem "rails", "3.2.22.2"
```
this is unrelated actually

marcoow commented on gemfiles/rails_4.gemfile:
```diff
@@ -2,7 +2,7 @@
 source "https://rubygems.org"
 
-gem "rails", "4.2.3"
+gem "rails", "4.2.6"
```
this is unrelated actually

marcoow commented on spec/spec_helper.rb:
```diff
@@ -2,8 +2,10 @@
+  unless defined?(JRUBY_VERSION) || defined?(JRuby)
+    minimum_coverage 95
+    refuse_coverage_drop
+  end
```
this is unrelated

marcoow commented:
Cool, there's still a few file mode changes left and also some unrelated changes.

dominicfollett commented:
Thanks I'll update and clean up those I missed.

dominicfollett commented:
@marcoow I think we're good to go? :)

marcoow commented on .rubocop.yml:
```diff
@@ -44,6 +44,9 @@ Style/Documentation:
 
+Metrics/MethodLength:
+  Enabled: false
+
```
Can you only disable this guard for the one method where it's failing?

dominicfollett replied:
I see, yeah that would be much better! Updated.

marcoow commented on app/controllers/oauth2_controller.rb:
```diff
@@ -7,6 +7,7 @@ class Oauth2Controller < ApplicationController
 
   force_ssl if: -> { RailsApiAuth.force_ssl }
 
+  # rubocop:disable all
```
Only the method length check should be disable and only for this specific method (enable it again after the method definition).

dominicfollett replied:
Thanks for your patience, I misunderstood the semantics of rubocop here.

dominicfollett commented:
@marcoow Thanks for your patience, and for pointing out those issues. I've re enabled rubocop method length checks after that create method.

marcoow commented:
🎉

marcoow commented on README.md:
```diff
@@ -156,16 +156,39 @@ RailsApiAuth.tap do |raa|
+  # Edx configurations
+  raa.edx_client_id     = '<your Edx client id>'
+  raa.edx_client_secret = '<your Edx client secret>'
+  raa.edx_domain        = '<your Edx app domain>'
```
Why do we need to customize this? Does Edx give every app its own API domain?
